File: v2/domain_api/librarian_mediation_inference_contract.py
"""Inference-facing contract projection for S2a librarian_mediation (#169)."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

_CONTRACT_PATH = (
    Path(__file__).resolve().parents[1]
    / "contracts"
    / "inference"
    / "hg_librarian_mediation_result_v1.inference-contract.json"
)


@lru_cache(maxsize=1)
def load_librarian_mediation_inference_contract() -> dict[str, Any]:
    with _CONTRACT_PATH.open(encoding="utf-8") as handle:
        return json.load(handle)


def _render_field_specs(field_specs: dict[str, Any]) -> list[str]:
    lines: list[str] = []
    for field_name, spec in field_specs.items():
        if spec.get("type") == "array":
            item_fields = spec.get("item_fields") or {}
            lines.append(f"- {field_name}: array")
            for item_field, item_spec in item_fields.items():
                req = "required" if item_spec.get("required") else "optional"
                extra = ""
                if item_spec.get("type") == "enum":
                    extra = f" one of {item_spec.get('values')}"
                elif item_spec.get("type") == "positive_integer":
                    extra = " positive integer"
                elif item_spec.get("catalog_only"):
                    extra = " catalog source_id only"
                lines.append(f"  - {item_field} ({req}){extra}")
        else:
            req = "required" if spec.get("required") else "optional"
            lines.append(f"- {field_name} ({req})")
    return lines


def render_mediation_inference_contract_lines(
    *,
    sample_source_id: str = "lmi:cand:example-source",
) -> str:
    contract = load_librarian_mediation_inference_contract()
    schema_id = contract["schema_id"]
    example = json.loads(json.dumps(contract["minimal_valid_example"]))
    if example.get("selected_items"):
        example["selected_items"][0]["source_id"] = sample_source_id
    if example.get("synthesis_entries"):
        example["synthesis_entries"][0]["source_ids"] = [sample_source_id]

    lines = [
        f"Required output contract for schema {schema_id}:",
        "",
        "Top-level object MUST include:",
        f'- schema: exact string "{schema_id}"',
        f"- required fields: {', '.join(contract.get('required_top_level', []))}",
        "",
        "Field specifications:",
        *_render_field_specs(contract.get("field_specs") or {}),
        "",
        "Forbidden field aliases (do NOT use):",
    ]
    for alias in contract.get("forbidden_field_aliases") or []:
        lines.append(f"- {alias.get('path')} — use {alias.get('use_instead')} instead")
    lines.append("")
    lines.append("Forbidden behaviors:")
    for rule in contract.get("forbidden_behaviors") or []:
        lines.append(f"- {rule}")
    lines.append("")
    lines.append("Minimal valid example (replace source_id with catalog values):")
    lines.append(json.dumps(example, ensure_ascii=False, indent=2))
    return "\n".join(lines)

File: v2/rp_runtime/src/lib/librarian-mediation-inference-contract.mjs
import contractJson from '../../../contracts/inference/hg_librarian_mediation_result_v1.inference-contract.json' with { type: 'json' };

import { LIBRARIAN_MEDIATION_RESULT_SCHEMA } from './librarian-mediation-envelope.mjs';

export const LIBRARIAN_MEDIATION_INFERENCE_CONTRACT = contractJson;

function renderFieldSpecs(fieldSpecs = {}) {
  const lines = [];
  for (const [fieldName, spec] of Object.entries(fieldSpecs)) {
    if (spec.type === 'array') {
      lines.push(`- ${fieldName}: array`);
      for (const [itemField, itemSpec] of Object.entries(spec.item_fields ?? {})) {
        const req = itemSpec.required ? 'required' : 'optional';
        let extra = '';
        if (itemSpec.type === 'enum') extra = ` one of ${JSON.stringify(itemSpec.values)}`;
        if (itemSpec.type === 'positive_integer') extra = ' positive integer';
        if (itemSpec.catalog_only) extra = ' catalog source_id only';
        lines.push(`  - ${itemField} (${req})${extra}`);
      }
    } else {
      const req = spec.required ? 'required' : 'optional';
      lines.push(`- ${fieldName} (${req})`);
    }
  }
  return lines;
}

export function buildLibrarianMediationContractSpec({
  sampleSourceId = 'lmi:cand:example-source',
} = {}) {
  const example = JSON.parse(JSON.stringify(contractJson.minimal_valid_example));
  if (example.selected_items?.[0]) {
    example.selected_items[0].source_id = sampleSourceId;
  }
  if (example.synthesis_entries?.[0]) {
    example.synthesis_entries[0].source_ids = [sampleSourceId];
  }
  return {
    schema: LIBRARIAN_MEDIATION_RESULT_SCHEMA,
    contract: contractJson,
    minimal_example: example,
    forbidden_aliases: contractJson.forbidden_field_aliases ?? [],
    forbidden_behaviors: contractJson.forbidden_behaviors ?? [],
  };
}

export function librarianMediationContractPromptLines({
  sampleSourceId = 'lmi:cand:example-source',
} = {}) {
  const spec = buildLibrarianMediationContractSpec({ sampleSourceId });
  const requiredTop = (contractJson.required_top_level ?? []).join(', ');
  const aliasLines = (spec.forbidden_aliases ?? []).map(
    (alias) => `- ${alias.path} — use ${alias.use_instead} instead`,
  );
  const behaviorLines = (spec.forbidden_behaviors ?? []).map((rule) => `- ${rule}`);
  return [
    `Required output contract for schema ${spec.schema}:`,
    '',
    'Top-level object MUST include:',
    `- schema: exact string "${spec.schema}"`,
    `- required fields: ${requiredTop}`,
    '',
    'Field specifications:',
    ...renderFieldSpecs(contractJson.field_specs ?? {}),
    '',
    'Forbidden field aliases (do NOT use):',
    ...aliasLines,
    '',
    'Forbidden behaviors:',
    ...behaviorLines,
    '',
    'Minimal valid example (replace source_id with catalog values):',
    JSON.stringify(spec.minimal_example, null, 2),
  ];
}

export function contractRequiredSelectedItemFields() {
  const itemFields = contractJson.field_specs?.selected_items?.item_fields ?? {};
  return Object.entries(itemFields)
    .filter(([, spec]) => spec.required)
    .map(([name]) => name);
}
